File: netlogo-gui/src/main/prim/_fdinternal.java
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim;

import org.nlogo.agent.Turtle;
import org.nlogo.api.AgentException;
import org.nlogo.core.Let;
import org.nlogo.core.Syntax;
import org.nlogo.nvm.Command;
import org.nlogo.nvm.Context;
import org.nlogo.nvm.MutableDouble;

// note that this and _bkinternal are pretty much carbon copies of each other

public final class _fdinternal
    extends Command {
  private final Let let;

  public _fdinternal(_fd original) {
    let = original.let;
    this.switches = true;
  }

  public _fdinternal(_bk original) {
    let = original.let;
    this.switches = true;
  }



  @Override
  public void perform(Context context) {
    perform_1(context);
  }

  public void perform_1(Context context) {
    Turtle turtle = (Turtle) context.agent;
    MutableDouble countdown = (MutableDouble) context.activation.binding.getLet(let);
    double distance = countdown.value();
    double distanceMagnitude = StrictMath.abs(distance);
    if (distanceMagnitude <= org.nlogo.api.Constants.Infinitesimal()) {
      context.ip = next;
      return;
    }
    if (distanceMagnitude <= 1.0) {
      try {
        turtle.jump(distance);
      } catch (AgentException e) { } // NOPMD
      context.ip = next;
    } else {
      int stepDistance = (distance > 0) ? 1 : -1;
      try {
        turtle.jump(stepDistance);
        countdown.value_$eq(countdown.value() - stepDistance);
      } catch (AgentException e) {
        context.ip = next;
      }
    }
  }
}

File: netlogo-headless/src/main/prim/_fdinternal.scala
// (C) Uri Wilensky. https://github.com/NetLogo/NetLogo

package org.nlogo.prim

import org.nlogo.agent.Turtle
import org.nlogo.api.{ AgentException, Numbers }
import org.nlogo.core.Let
import org.nlogo.nvm.{ Command, Context, MutableDouble }

class _fdinternal(let: Let) extends Command {

  def this(original: _fd) = this(original.let)
  def this(original: _bk) = this(original.let)

  switches = true

  override def perform(context: Context): Unit = {
    perform_1(context)
  }

  def perform_1(context: Context): Unit = {
    val turtle = context.agent.asInstanceOf[Turtle]
    val countdown = context.activation.binding.getLet(let).asInstanceOf[MutableDouble]
    val distance = countdown.value
    val distanceMagnitude = StrictMath.abs(distance)
    if (distanceMagnitude <= Numbers.Infinitesimal)
      context.ip = next
    else if (distanceMagnitude <= 1.0) {
      try turtle.jump(distance)
      catch { case _: AgentException => }
      context.ip = next
    } else {
      val stepDistance = if (distance > 0) 1 else -1
      try {
        turtle.jump(stepDistance)
        countdown.value -= stepDistance
      }
      catch { case _: AgentException =>
        context.ip = next
      }
    }
  }
}
